File: communicationAPI/src/main/java/eu/cybergeiger/communication/StorableHashMap.java
package eu.cybergeiger.communication;

import ch.fhnw.geiger.serialization.Serializer;
import ch.fhnw.geiger.serialization.SerializerHelper;
import ch.fhnw.geiger.totalcross.ByteArrayInputStream;
import ch.fhnw.geiger.totalcross.ByteArrayOutputStream;
import ch.fhnw.geiger.totalcross.System;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * Serializable Hashmap.
 *
 * @param <K> Keytype
 * @param <V> Valuetype
 */
public class StorableHashMap<K extends Serializer, V extends Serializer>
    extends HashMap<K, V> implements Serializer {

  private static final long serialVersionUID = 14231491232L;

  @Override
  public void toByteArrayStream(ByteArrayOutputStream out) throws IOException {
    SerializerHelper.writeLong(out, serialVersionUID);
    SerializerHelper.writeInt(out, size());
    for (Map.Entry e : entrySet()) {
      SerializerHelper.writeObject(out, e.getKey());
      SerializerHelper.writeObject(out, e.getValue());
    }
    SerializerHelper.writeLong(out, serialVersionUID);
  }

  /**
   * <p>Reads objects from ByteArrayInputStream and stores them in map.</p>
   *
   * @param in  ByteArrayInputStream to be used
   * @param map Map to store objects
   * @throws IOException if value cannot be read
   */
  public static void fromByteArrayStream(ByteArrayInputStream in, StorableHashMap map)
      throws IOException {
    if (SerializerHelper.readLong(in) != serialVersionUID) {
      throw new ClassCastException();
    }
    map.clear();
    int size = SerializerHelper.readInt(in);
    for (int i = 0; i < size; i++) {
      map.put(SerializerHelper.readObject(in), SerializerHelper.readObject(in));
    }
    if (SerializerHelper.readLong(in) != serialVersionUID) {
      throw new ClassCastException();
    }
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    for (Map.Entry e : entrySet()) {
      sb.append(e.getKey().toString()).append('=').append(e.getValue())
          .append(System.lineSeparator());
    }
    return sb.toString();
  }
}

File: lib/src/communication/storable_hash_map.dart
library geiger_api;

import 'package:geiger_api/src/serialization/communication_serializer.dart';
import 'package:geiger_localstorage/geiger_localstorage.dart';

// TODO: Maybe use quiver's DelegatingMap:
// https://pub.dev/documentation/quiver/latest/quiver.collection/DelegatingMap-class.html

/// Serializable Hashmap.
class StorableHashMap<K extends Serializer, V extends Serializer>
    with Serializer
    implements Map<K, V> {
  static const int serialVersionUID = 14231491232;

  final Map<K, V> _map = <K, V>{};

  @override
  void toByteArrayStream(ByteSink out) {
    SerializerHelper.writeLong(out, serialVersionUID);
    SerializerHelper.writeInt(out, length);
    for (MapEntry<K, V> e in entries) {
      writeObject(out, e.key);
      writeObject(out, e.value);
    }
    SerializerHelper.writeLong(out, serialVersionUID);
  }

  /// Reads objects from ByteArrayInputStream and stores them in map.
  ///
  /// @param in  ByteArrayInputStream to be used
  /// @param map Map to store objects
  /// @throws IOException if value cannot be read
  static Future<StorableHashMap> fromByteArrayStream(
      ByteStream in_, StorableHashMap map) async {
    if (await SerializerHelper.readLong(in_) != serialVersionUID) {
      throw Exception('cannot cast');
    }
    map.clear();
    int size = await SerializerHelper.readInt(in_);
    for (int i = 0; i < size; i++) {
      map[(await readObject(in_))!] = (await readObject(in_))!;
    }
    if (await SerializerHelper.readLong(in_) != serialVersionUID) {
      throw Exception('cannot cast');
    }
    return map;
  }

  @override
  String toString() {
    StringBuffer sb = StringBuffer();
    for (MapEntry e in entries) {
      sb.write(e.key.toString());
      sb.write('=');
      sb.write(e.value);
      sb.write('\n');
    }
    return sb.toString();
  }

  @override
  V? operator [](Object? key) {
    return _map[key];
  }

  @override
  void operator []=(K key, V value) {
    _map[key] = value;
  }

  @override
  void addAll(Map<K, V> other) {
    _map.addAll(other);
  }

  @override
  void addEntries(Iterable<MapEntry<K, V>> newEntries) {
    _map.addEntries(newEntries);
  }

  @override
  Map<RK, RV> cast<RK, RV>() {
    return _map.cast<RK, RV>();
  }

  @override
  void clear() {
    _map.clear();
  }

  @override
  bool containsKey(Object? key) {
    return _map.containsKey(key);
  }

  @override
  bool containsValue(Object? value) {
    return _map.containsValue(value);
  }

  @override
  Iterable<MapEntry<K, V>> get entries {
    return _map.entries;
  }

  @override
  void forEach(void Function(K key, V value) action) {
    _map.forEach(action);
  }

  @override
  bool get isEmpty {
    return _map.isEmpty;
  }

  @override
  bool get isNotEmpty {
    return _map.isNotEmpty;
  }

  @override
  Iterable<K> get keys {
    return _map.keys;
  }

  @override
  int get length {
    return _map.length;
  }

  @override
  Map<K2, V2> map<K2, V2>(MapEntry<K2, V2> Function(K key, V value) convert) {
    return _map.map(convert);
  }

  @override
  V putIfAbsent(K key, V Function() ifAbsent) {
    return _map.putIfAbsent(key, ifAbsent);
  }

  @override
  V? remove(Object? key) {
    return _map.remove(key);
  }

  @override
  void removeWhere(bool Function(K key, V value) test) {
    return _map.removeWhere(test);
  }

  @override
  void updateAll(V Function(K key, V value) update) {
    _map.updateAll(update);
  }

  @override
  Iterable<V> get values {
    return _map.values;
  }

  @override
  V update(K key, V Function(V value) update, {V Function()? ifAbsent}) {
    return _map.update(key, update, ifAbsent: ifAbsent);
  }
}
